monitor/monitor.go

```diff
@@ -35,82 +36,157 @@ const (
 		}
+
+		m.send(&p)
 	}
 }
 
-// Run starts monitoring.
-func (m *Monitor) Run(npages int, agentPipe io.Reader) {
-	stopAgentPipeReader := make(chan struct{})
-	go m.agentPipeReader(agentPipe, stopAgentPipeReader)
-	defer close(stopAgentPipeReader)
-
+// NewMonitor creates a Monitor, and starts client connection handling and agent event
+// handling.
+// Note that the perf buffer reader is started only when listeners are
+// connected.
+func NewMonitor(ctx context.Context, npages int, agentPipe io.Reader, server net.Listener) (m *Monitor, err error) {
+	// configure BPF perf buffer reader
 	c := bpf.DefaultPerfEventConfig()
 	c.NumPages = npages
 
-	me, err := bpf.NewPerCpuEvents(c)
+	monEvts, err := bpf.NewPerCpuEvents(c)
 	if err != nil {
-		log.WithError(err).Error("Error while starting monitor")
-		return
+		return nil, err
+	}
+
+	m = &Monitor{
+		ctx:           ctx,
+		listeners:     make(map[*monitorListener]struct{}),
+		monitorEvents: monEvts,
 	}
-	monitorEvents = me
+
+	// start new listener handler
+	go m.connectionHandler(ctx, server)
+
+	// start agent event pipe reader
+	go m.agentPipeReader(ctx, agentPipe)
+
+	return m, nil
+}
+
+// registerNewListener adds the new listener to the global list. It also spawns
+// a singleton goroutine to read and distribute the events. It passes a
+// cancelable context to this goroutine and the cancelFunc is assigned to
+// perfReaderCancel. Note that cancelling parentCtx (e.g. on program shutdown)
+// will also cancel the derived context.
+func (m *Monitor) registerNewListener(parentCtx context.Context, conn net.Conn) {
+	m.Lock()
+	defer m.Unlock()
+
+	// If this is the first listener, start the perf reader
+	if len(m.listeners) == 0 {
+		perfEventReaderCtx, cancel := context.WithCancel(parentCtx)
+		m.perfReaderCancel = cancel
+		go m.perfEventReader(perfEventReaderCtx)
+	}
+
+	newListener := newMonitorListener(conn, m.removeListener)
+	m.listeners[newListener] = struct{}{}
+
+	log.WithField("count.listener", len(m.listeners)).Info("New listener connected.")
+}
+
+// removeListener deletes the listener from the list, closes it's queue, and
+// stops perfReader if this is the last listener
+func (m *Monitor) removeListener(ml *monitorListener) {
+	m.Lock()
+	defer m.Unlock()
+
+	delete(m.listeners, ml)
+	log.WithField("count.listener", len(m.listeners)).Info("Removed listener")
+
+	// If this was the final listener, shutdown the perf reader
+	if len(m.listeners) == 0 {
+		m.perfReaderCancel()
+	}
+}
+
+func (m *Monitor) perfEventReader(stopCtx context.Context) {
+	log.Info("Beginning to read perf buffer")
+	defer log.Info("Stopped reading perf buffer")
 
 	last := time.Now()
-	// Main event loop
-	for {
-		todo, err := monitorEvents.Poll(pollTimeout)
-		if err != nil {
+	for !isCtxDone(stopCtx) {
+		todo, err := m.monitorEvents.Poll(pollTimeout)
+		switch {
+		case isCtxDone(stopCtx):
+			return
+
+		case err == syscall.EBADF:
```

Comment:
Should this also print the "Error in Poll" below? I'm not sure why we handled this case differently